src/geo/geohash.clj

```diff
@@ -192,82 +188,38 @@
-              (take-while not-empty)
-              (apply concat))))))
-
-(defn geohashes-intersecting-recursive
-  "Starting with the given geohash which completely encloses the target shape,
-  recursively subdivides and filters to compute a set of geohashes of the given
-  resolution which intersect that shape."
-  [shape precision ^GeoHash geohash]
-  (let [relationship (relate shape geohash)]
-    (cond
-      (= relationship :contains)
-      (do
-;        (print "X")
-        (subdivide geohash precision))
+(defn children [gh]
+  (->> gh string GeohashUtils/getSubGeohashes (map geohash)))
 
-      (not= relationship :disjoint)
-      (let [current-precision (.significantBits geohash)
-            delta (- precision current-precision)]
-;        (print ".")
-        (if (zero? delta)
-          ; Done
-          (list geohash)
-          ; Keep going
-          (mapcat (partial geohashes-intersecting-recursive shape precision)
-                  ; Split each geohash into quads (or halves if necessary.
-                  (subdivide geohash (+ (.significantBits geohash)
-                                        (min 2 delta)))))))))
+(defn bbox-geom [geohash]
+  (let [bbox (.getBoundingBox geohash)
+        rect (RectangleImpl. (.getMinLon bbox)
+                             (.getMaxLon bbox)
+                             (.getMinLat bbox)
+                             (.getMaxLat bbox)
+                             JtsSpatialContext/GEO)]
+    (.getGeometryFrom JtsSpatialContext/GEO rect)))
 
 (defn geohashes-intersecting
-  "A hybrid algorithm to find all the geohashes of a given resolution which
-  intersect a given shape. Finds an initial set of hashes with concentric
-  rings, then refines those hashes by subdividing them. If given,
-  initial-precision specifies the size of the geohashes used with the slow,
-  concentric-rings algorithm to identify geohashes to subdivide."
-  ([shape precision]
-   (let [initial-precision (max 0 (- (shape->precision shape) 6))]
-;     (println "Starting with precision " initial-precision " ("
-;              (nth degrees-precision-lat-cache  initial-precision) " x "
-;              (nth degrees-precision-long-cache initial-precision) 
-;              ") enclosing "
-;              (height shape) " x " (width shape))
-     (geohashes-intersecting shape precision initial-precision)))
-  ([shape precision initial-precision]
-   (mapcat (partial geohashes-intersecting-recursive shape precision)
-           (geohashes-intersecting-rings shape initial-precision))))
+  ([shape desired-level] (geohashes-intersecting shape desired-level desired-level))
+  ([shape min-level max-level]
+   (loop [matches #{}
+          queue (list (geohash ""))]
+     (if (empty? queue) matches
+         (let [current (first queue)
+               level (significant-bits current)]
+
+           (if (and (<= level max-level) (intersects? shape current))
+             (if (= level max-level) (recur (conj matches current)
```

Comment:
mixing same-line and multi-line if-elses is very hard to read, you generally want to use one or the other. I.e. do either

```clojure
(if my-cond? 1 0)
```
or 
```clojure
(if my-cond?
  1
  0)
```
But mixing is very confusing:
```clojure
(if my-cond? 1
  0)
```